cpp/include/cugraph/device_vector.hpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph_c/types.h>
+
+#include <cugraph/utilities/cugraph_data_type_id.hpp>
+#include <cugraph/utilities/error.hpp>
+#include <cugraph/utilities/packed_bool_utils.hpp>
+
+#include <raft/core/handle.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+
+namespace cugraph {
+
+namespace detail {
+
+inline rmm::device_buffer allocate_buffer(raft::handle_t const& handle,
+                                          cugraph_data_type_id_t t,
+                                          size_t size)
+{
+  return (t == BOOL) ? rmm::device_buffer(cugraph::packed_bool_size(size) * data_type_size(t),
+                                          handle.get_stream())
+                     : rmm::device_buffer(size * data_type_size(t), handle.get_stream());
+}
+
+}  // namespace detail
+
+/**
+ * Class wrapping a type-erased device vector.
+ * */
+class device_vector_t {
```

Comment:
I am not sure `device_vector` is a right name for type erased edge property vector. `device_vector` reminds me of `thrust::device_vector` or `rmm::device_uvector`. They are not type-erased vectors.

We may need a better name. My suggestions: `edge_property_buffer` or `property_buffer` or `cugraph::device_buffer`. `rmm::device_buffer` does not assume a specific type and may be more appropriate for type erased data types. Or we may use `column` (adopt column from `cudf::column`). In cuDF C++, they use `cudf::column` to store cuDF series. We may say `edge_property_column` or `property_column` or `cugraph::device_column` or just `cugraph::column`. In this case, we may rename the above `allocate_buffer` function to `allocate_column_buffer` as well.

Reply:
I feel like we need more discussion about this naming.